presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java

```diff
@@ -675,6 +706,11 @@ private int calculateNextBatchSize(int fixedWidthRowSize, List<Integer> variable
                 }
             }
 
+            // Return the borrowed array for serializedRowSizes when the current page will be finished in this batch.
+            // At this point the batchSize was already calculated and we don't need the serializedRowSizes any more.
+            allocator.returnArray(serializedRowSizes);
```

Comment:
This logic is confusing to me. calculateNextBatchSize is called in a do-while loop and therefore it is not clear that it won't be called anymore. I think it would be clearer to move allocation and freeing of the serializedRowSizes into appendData and change serializedRowSizes into a local variable:

```
            int[] serializedRowSizes = ensureCapacityWithAllocator(null, positionCount, SMALL, INITIALIZE, allocator);
            try {

                populateSerializedRowSizes(fixedWidthRowSize, variableWidthChannels, serializedRowSizes);

                // Due to the limitation of buffer size, we append the data batch by batch
                int offset = 0;
                do {
                    int batchSize = calculateNextBatchSize(fixedWidthRowSize, variableWidthChannels, offset, serializedRowSizes);
                    boolean lastBatchInPage = offset + batchSize >= positionCount;

                    for (int i = 0; i < channelCount; i++) {
                        blockEncodingBuffers[i].setNextBatch(offset, batchSize);
                        blockEncodingBuffers[i].appendDataInBatch(lastBatchInPage);
                    }

                    bufferedRowCount += batchSize;
                    offset += batchSize;

                    if (bufferFull) {
                        flush(outputBuffer);
                        bufferFull = false;
                    }
                }
                while (offset < positionCount);
            }
            finally {
                // Return the borrowed array for serializedRowSizes when the current page will be finished in this batch.
                // At this point the batchSize was already calculated and we don't need the serializedRowSizes any more.
                allocator.returnArray(serializedRowSizes);
            }
```

Reply:
@mbasmanova The original logic was to release the serializedRowSizes after the size of the last batch of the page was calculated. For that purpose the returnArray() happens before the last line of calculateNextBatchSize(). If it's not the last batch in page, the code would have returned before reaching that point. I was doing it this way because I thought serializedRowSizes is an internal thing only used for calculating batchSize, therefore it was better to be confined inside of populateSerializedRowSizes() and calculateNextBatchSize(). As the caller of these two methods, appendData() doesn't need to care about how the serializedRowSizes is used. What's your opinion on this? If you think this is not important, I'll change it the way you suggested. Thanks!

Reply:
Masha, I updated the code as you suggested.